[CLOUDCRAFT-388] Automate Snapshots of Cloud Accounts via the Cloudcraft API (#22262)
DataDog/documentation: content/en/cloudcraft/advanced/auto-layout-via-api.md

---
title: Automate Snapshots of Cloud Accounts via the Cloudcraft API
kind: guide
---

## Overview

Cloudcraft's **Auto Layout** feature, accessible through the web application, is a powerful tool for automatically generating diagrams of your AWS environment. This functionality can significantly streamline documentation processes and facilitate the onboarding of new team members.

This guide provides a step-by-step approach to utilizing this feature via common command-line utilities and the Cloudcraft developer API.

> <div class="alert alert-info">The ability to add and scan AWS accounts, as well as to use Cloudcraft's developer API, is only available to Pro subscribers. Check out <a href="https://www.cloudcraft.co/pricing">Cloudcraft's pricing page</a> for more information.</div>

Author: Fixed with [8cb0721](https://github.com/DataDog/documentation/pull/22262/commits/8cb0721724ea3ad5c892c35119d94c85f74c619f).

## Prerequisites

- An active [Cloudcraft Pro subscription][1].
- An API key with read-write permissions.
- Access to a Unix-like environment (Linux, macOS, or Windows Subsystem for Linux).
- Familiarity with command-line operations.
- Basic knowledge of API usage.

> ## Take a snapshot of the account

Reviewer: Do we need to add a step here to get the AWS / Azure account id using list accounts API and identify the right value for ACCOUNT_ID?

Since these ACCOUNT_IDs generated by Cloudcraft, the user will not know this offhand.

Author: I think that should be its own article, which we need anyway. This come up on support all the time because of access control.

Reviewer: Should we alteast then mention that in the pre-requisites - you should have the Cloudcraft issued identifier for the interested AWS account handy?

Author: Added on [c672e98](https://github.com/DataDog/documentation/pull/22262/commits/c672e98c3afa08a4c355494a46ef046643ef83cc).

Start by creating a snapshot of your AWS or Azure account using the [Snapshot AWS account][2] or [Snapshot Azure account][3] endpoints. This process mirrors the functionality of the **Scan Now** button in the Cloudcraft UI and outputs the snapshot in JSON format.

Execute the following command in your terminal:

> **For AWS:**

Reviewer: Instead of calling out separately for AWS and Azure , you can also make that a parameter like PROVIDER where provider value is `aws` or `azure`

Author: Oh, great idea! Doing that now.

Author: Fixed with [685aacc](https://github.com/DataDog/documentation/pull/22262/commits/685aacc3b5f5d89410ea3802276ae65ce1f19b4a).

{{< code-block lang="shell" >}}
curl \
  --url 'https://api.cloudcraft.co/aws/account/ACCOUNT_ID/REGION/json' \
  --tlsv1.2 \
  --proto '=https' \
  --silent \
  --header "Authorization: Bearer API_KEY"
{{< /code-block >}}

**For Azure:**

{{< code-block lang="shell" >}}
curl \
  --url 'https://api.cloudcraft.co/azure/account/ACCOUNT_ID/REGION/json' \
  --tlsv1.2 \
  --proto '=https' \
  --silent \
  --header "Authorization: Bearer API_KEY"
{{< /code-block >}}

Replace ACCOUNT_ID with the ID of your AWS or Azure account in Cloudcraft, REGION with your desired scan region, and API_KEY with your Cloudcraft API key.

After executing the command, you will see the JSON representation of your AWS account snapshot. To save this output directly to a file, use the following command:

> **For AWS:**

Reviewer: Instead of calling out separately for AWS and Azure , you can also make that a parameter like PROVIDER where provider value is `aws` or `azure`

Author: Fixed with [685aacc](https://github.com/DataDog/documentation/pull/22262/commits/685aacc3b5f5d89410ea3802276ae65ce1f19b4a).

<div class="alert alert-info">The ability to add and scan AWS and Azure accounts, as well as to use Cloudcraft's developer API, is only available to Pro subscribers. Check out <a href="https://www.cloudcraft.co/pricing">Cloudcraft's pricing page</a> for more information.</div>

## Prerequisites

- An active [Cloudcraft Pro subscription][1].
- An API key with read-write permissions.
- The account ID of the AWS or Azure account you wish to scan.
- Access to a Unix-like environment (Linux, macOS, or Windows Subsystem for Linux).
- Familiarity with command-line operations.
- Basic knowledge of API usage.

## Take a snapshot of the account

Start by creating a snapshot of your AWS or Azure account using the [Snapshot AWS account][2] or [Snapshot Azure account][3] endpoints. This process mirrors the functionality of the **Scan Now** button in the Cloudcraft UI and outputs the snapshot in JSON format.

Execute the following command in your terminal:

{{< code-block lang="shell" >}}
curl \
  --url 'https://api.cloudcraft.co/PROVIDER/account/ACCOUNT_ID/REGION/json' \
  --tlsv1.2 \
  --proto '=https' \
  --silent \
  --header "Authorization: Bearer API_KEY"
{{< /code-block >}}

> Replace PROVIDER with the cloud provider, i.e., `azure` or `aws`, ACCOUNT_ID with the ID of your AWS or Azure account in Cloudcraft, REGION with your desired scan region, and API_KEY with your Cloudcraft API key.

Reviewer 2: ```suggestion
Replace `PROVIDER` with the cloud provider, for example, `azure` or `aws`, `ACCOUNT_ID` with the ID of your AWS or Azure account in Cloudcraft, `REGION` with your desired scan region, and `API_KEY` with your Cloudcraft API key.
```

> After executing the command, you will see the JSON representation of your AWS account snapshot. To save this output directly to a file, use the following command:

Reviewer 2: ```suggestion
After executing the command, the JSON representation of your AWS account snapshot is displayed. To save this output directly to a file, use the following command:
```

{{< code-block lang="shell" >}}
curl \
  --url 'https://api.cloudcraft.co/PROVIDER/account/ACCOUNT_ID/REGION/json' \
  --tlsv1.2 \
  --proto '=https' \
  --silent \
  --header "Authorization: Bearer API_KEY" > '/tmp/account-infra.json'
{{< /code-block >}}

> You will find the snapshot saved as `account-infra.json` in your temporary directory.

Reviewer 2: ```suggestion
The snapshot is saved with the filename `account-infra.json` in your temporary directory.
```

## Generate a new blueprint

> With the account snapshot in hand, proceed to create a new blueprint in your Cloudcraft account using the [Create blueprint][4] API endpoint. The saved snapshot data will serve as the payload for this request.

Reviewer 2: ```suggestion
Next, create a new blueprint in your Cloudcraft account using the [Create blueprint][4] API endpoint. The saved snapshot data serves as the payload for this request.
```

Execute the following command in your terminal:

{{< code-block lang="shell" >}}
curl \
  --request 'POST' \
  --url 'https://api.cloudcraft.co/blueprint' \
  --tlsv1.2 \
  --proto '=https' \
  --silent \
  --header 'Content-Type: application/json' \
  --header "Authorization: Bearer API_KEY" \
  --data '@/tmp/account-infra.json'
{{< /code-block >}}

> Remember to replace API_KEY with your actual Cloudcraft API key.

Reviewer 2: ```suggestion
Remember to replace `API_KEY` with your actual Cloudcraft API key.
```

> Upon completion, a new blueprint reflecting your cloud infrastructure will be created in your Cloudcraft account, replicating the effect of manually using the **Scan Now** and **Auto Layout** buttons.

Reviewer 2: ```suggestion
Upon completion, a new blueprint reflecting your cloud infrastructure is created in your Cloudcraft account, replicating the effect of manually using the **Scan Now** and **Auto Layout** buttons.
```

> If you have any questions or trouble with the process, [get in touch with our support team][5] and we will be happy to help.

Reviewer 2: ```suggestion
If you have any questions or trouble with the process, [contact Cloudcraft's support team][5].
```
